src/evm/protocol/uniswap_v4/angstrom/streams.rs

```diff
@@ -0,0 +1,168 @@
+            return Err(StreamError::SetUpError("Angstrom url requires websocket.".to_string()));
+        }
+
+        let (receiving_tx, receiving_rx) = channel(100);
+        let (update_sender, sending_rx) = channel(100);
+        let update_receiver = ReceiverStream::new(receiving_rx);
+
+        let channels = ProtocolStreamExtension { tx: Some(receiving_tx), rx: sending_rx };
+
+        let client = WsClientBuilder::default()
+            .build(angstrom_url)
+            .await
+            .map_err(|e| StreamError::WebSocketConnectionError(e.to_string()))?;
+        let arced_client = Arc::new(client);
+        // connect to angstrom streams.
+        let slot0_client = Slot0Client::new(arced_client.clone());
+        let unlock_client = UnlockClient::new(arced_client);
+
+        let this = Self {
+            slot0_client,
+            unlock_client,
+            tycho_header: Header::default(),
+            update_receiver,
+            update_sender,
+        };
+        tokio::spawn(this.run());
+
+        Ok(channels)
+    }
+
+    pub async fn send_slot0_update(&mut self, update: Slot0Update) {
+        if update.current_block != self.tycho_header.number {
+            return;
+        }
+
+        let mut attributes = HashMap::new();
+
+        attributes.insert("liquidity".to_string(), Bytes::from(update.liquidity.to_be_bytes()));
+        attributes.insert(
+            "sqrt_price_x96".to_string(),
+            Bytes::from(U256::from(update.sqrt_price_x96).to_be_bytes::<32>()),
+        );
+        attributes.insert(
+            "tick".to_string(),
+            Bytes::from(I24::unchecked_from(update.tick).to_be_bytes::<3>()),
+        );
+
+        let state_update = ProtocolStateDelta {
+            component_id: update.uni_pool_id.to_string(),
+            updated_attributes: attributes,
+            ..Default::default()
+        };
+
+        let mut state_updates = HashMap::new();
+        state_updates.insert("angstrom".to_string(), state_update);
+
+        let deltas = BlockChanges { state_updates, ..Default::default() };
+        let message = StateSyncMessage {
+            header: self.tycho_header.clone(),
+            deltas: Some(deltas),
+            ..Default::default()
+        };
+
+        let state_msgs = HashMap::from_iter([("angstrom".to_string(), message)]);
+
+        let message = FeedMessage { state_msgs, sync_states: Default::default() };
+        self.update_sender
+            .send(message)
+            .await
+            .expect("feed is no longer being serviced");
+    }
+    pub async fn send_unlock_update(&mut self, update: ConsensusDataWithBlock<Bytes>) {
+        todo!()
+    }
+
+    pub async fn on_tycho_message(&mut self, feed_message: Arc<FeedMessage>) {
+        if let Some(state_message) = feed_message.state_msgs.get("angstrom") {
+            self.tycho_header = state_message.header.clone();
+        }
+    }
```

Comment:
I think you want to query `"uniswap-v4"` here, not `"angstrom"`.

But the main issue here is that we potentially want to undo previously sent Slot0 updates in case they are not included in the block or is this absolutely impossible to happen? I am ok to ignore for now but might leave the pool in a potential "stale" state.

Reply:
So slot0 updates will stream valid pool state in the case that there are no orders, because of this, the only thing that needs to be added is detecting if the slot0 stream stops sending the 100ms updates (will attempt to reconnect) and then we want to rollback to the lastest state to avoid staleness, However If slot0 stream is working fine, It will just stream the correct state itself so we don't have to worry about what the tycho-indexer is sending.